Fix for cbmc running out of memory while printing traces using json_ui / target: develop

NlightNFotis opened the pull request:
This is a port of #1767 to `develop` (instead of `goto-analyzer-develop` as requested by @peterschrammel and @martin-cs . Below is the message of the original PR:

There has been a bug reported to us where cbmc would run out of memory while creating a huge json object in memory to report the traces with. This is changing the tracing report facility to use a new json class that streams the objects immediately as created, and therefore fixes the issue.

peterschrammel commented on src/goto-programs/json_goto_trace.cpp:
```diff
@@ -11,4 +11,172 @@ Author: Daniel Kroening
 /// \file
 /// Traces of GOTO Programs
 
+#include <cassert>
```
This can be removed now.

peterschrammel replied:
The commit message isn't very useful.

peterschrammel commented on src/util/ui_message.h:
```diff
@@ -61,14 +61,6 @@ class ui_message_handlert:public message_handlert
     int sequence_number,
     const source_locationt &location) override;
 
-  virtual void print(
```
This change is in the wrong commit.

peterschrammel commented on src/cbmc/all_properties.cpp:
```diff
@@ -206,24 +206,23 @@ void bmc_all_propertiest::report(const cover_goalst &cover_goals)
         {
-          jsont &json_trace=result["trace"];
-          convert(bmc.ns, g.second.goto_trace, json_trace, bmc.trace_options());
+          json_stream_arrayt &json_trace=result.push_back_stream_array("trace");
```
(commit for entire commit) This commit should come after the json_goto_trace refactoring.

peterschrammel commented on src/util/json.cpp:
```diff
@@ -133,6 +117,32 @@ void jsont::output_rec(std::ostream &out, unsigned indent) const
   }
 }
 
+void jsont::output_object(std::ostream &out, const objectt &object, unsigned indent)
```
Some doxygen would be good (sorry that I threw that over to you without docs).

peterschrammel commented on src/util/json_stream.h:
```diff
@@ -10,6 +10,8 @@ Author: Peter Schrammel
 #ifndef CPROVER_UTIL_JSON_STREAM_H
 #define CPROVER_UTIL_JSON_STREAM_H
 
+#include <memory>
```
this should go into the commit where this file was introduced

peterschrammel replied:
these changes should go into the first commit

peterschrammel commented on src/util/ui_message.cpp:
```diff
@@ -123,51 +122,6 @@ void ui_message_handlert::print(
   }
 }
 
-void ui_message_handlert::print(
```
This seems still necessary as XML output hasn't been moved an xml_stream yet.

peterschrammel commented on src/util/ui_message.cpp:
```diff
@@ -6,8 +6,6 @@ Author: Daniel Kroening, kroening@kroening.com
 
 \*******************************************************************/
 
-#include "ui_message.h"
```
This include should come first.

peterschrammel commented on src/util/ui_message.cpp:
```diff
@@ -17,6 +15,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "json_expr.h"
 #include "json_stream.h"
 #include "cout_message.h"
+#include "ui_message.h"
```
Unclear why this file is modified in this commit.

peterschrammel commented on src/goto-programs/json_goto_trace.cpp:
```diff
@@ -44,27 +45,59 @@ void convert_assert(json_objectt &json_failure,
 
-  if(!location.is_null())
-    json_assignment["sourceLocation"]=location;
+  if(!json_location.is_null())
```
(comment for entire commit) These changes should be squashed with the previous commit.

peterschrammel replied:
I suggest to improve the commit message.

peterschrammel commented on src/util/json.h:
```diff
@@ -109,7 +109,6 @@ class jsont
 protected:
   void output_rec(std::ostream &, unsigned indent) const;
   static void escape_string(const std::string &, std::ostream &);
-
```
unclear why this line is removed

peterschrammel commented on src/util/json_stream.cpp:
```diff
@@ -0,0 +1,123 @@
+
+#include <ostream>
+
+void json_streamt::output_delimiter()
```
there should be some doxygen for all this

peterschrammel commented on src/util/json_stream.h:
```diff
@@ -0,0 +1,129 @@
+class json_streamt
+{
+public:
+  void close()
```
doxygen

peterschrammel commented on src/util/json_stream.cpp:
```diff
@@ -52,13 +52,15 @@ json_stream_objectt::json_stream_objectt(std::ostream &out, unsigned indent)
   out << '{';
 }
 
+/// Create a new JSON array stream.
```
The changes to this file should be in the previous commit.

peterschrammel commented on src/util/message.h:
```diff
@@ -232,6 +238,12 @@ class messaget
       return func(*this);
     }
 
+    json_stream_arrayt &json_stream()
```
doxygen

tautschnig replied:
Why is there a `json_stream_arrayt` variant, but no `json_stream_objectt` one?

peterschrammel commented on src/goto-programs/json_goto_trace.cpp:
```diff
@@ -11,279 +11,244 @@ Author: Daniel Kroening
-  json_arrayt &dest_array=dest.make_array();
 
-  source_locationt previous_source_location;
+#include "json_goto_trace.h"
```
this include should be the first

peterschrammel commented on src/goto-programs/json_goto_trace.h:
```diff
@@ -15,13 +15,123 @@ Date: November 2005
+
+typedef struct
+{
+  jsont &location;
```
why is this non-const?

peterschrammel commented on src/goto-programs/json_goto_trace.cpp:
```diff
@@ -11,279 +11,244 @@ Author: Daniel Kroening
-  for(const auto &step : goto_trace.steps)
-  {
-    const source_locationt &source_location=step.pc->source_location;
+void convert_assert(json_objectt &json_failure,
```
doxygen missing throughout these files

peterschrammel commented on src/goto-programs/json_goto_trace.cpp:
```diff
@@ -11,279 +11,244 @@ Author: Daniel Kroening
+ conversion_dependenciest &conversion_dependencies)
+{
+  const goto_trace_stept& step=conversion_dependencies.step;
+  jsont &location=conversion_dependencies.location;
```
why is this non-const?

peterschrammel commented:
@NlightNFotis, there's a compiliation error: https://travis-ci.org/diffblue/cbmc/jobs/339128397#L1550

peterschrammel commented:
Also, cpplint and clang-format are complaining. Be careful to reformat the changes in each commit separately.

peterschrammel commented on src/goto-programs/json_goto_trace.h:
```diff
@@ -15,13 +15,123 @@ Date: November 2005
+
+#include <util/json_expr.h>
+#include <util/arith_tools.h>
+#include <util/config.h>
```
There are a couple of header files that can be put into the cpp file.

NlightNFotis commented:
This should be ready to review, this only needs the clang format changes to be squashed to their appropriate commits.

tautschnig commented on src/util/json.cpp:
```diff
@@ -133,6 +120,41 @@ void jsont::output_rec(std::ostream &out, unsigned indent) const
+    if(o_it!=object.begin())
+      out << ',';
+
+    out << '\n';
```
Shouldn't the newline also be conditional on `o_it!=object.beginn()`?

peterschrammel replied:
No, because an object always starts with an opening brace, after which we put a newline (code snippet factored out from lines 69-85 above).

tautschnig commented on src/util/json.cpp:
```diff
@@ -133,6 +120,41 @@ void jsont::output_rec(std::ostream &out, unsigned indent) const
+  out << '"';
+  jsont::escape_string(key, out);
+  out << '"';
+  out << ": ";
```
Not a big deal, but `out << "\": ";` would also do the trick, wouldn't it?

tautschnig commented on src/util/json.h:
```diff
@@ -127,8 +127,14 @@ class jsont
   std::string value;
+
+  friend class json_stream_objectt;
+  friend class json_stream_arrayt;
```
Remote friends are bad.

peterschrammel replied:
Seems unnecessary to have them.

tautschnig commented on src/util/json_stream.cpp:
```diff
@@ -0,0 +1,142 @@
+  {
+    current->close();
+  }
+}
```
To be honest, that's a very strange piece of code. It seems to be extremely context-dependent and far from side-effect free.

tautschnig commented on src/util/json_stream.cpp:
```diff
@@ -0,0 +1,142 @@
+{
+  out << '\n' << std::string(indent*2, ' ');
+  out << ']';
+}
```
The constructor does `out << '['` so I'd expect this to be done in a destructor (otherwise the constructor shouldn't be doing the opening bracket).

peterschrammel replied:
This is exactly what it is doing (by calling close() which, in turn, calls output_finalizer).

tautschnig commented on src/util/json_stream.cpp:
```diff
@@ -0,0 +1,142 @@
+  current =
+    std::unique_ptr<json_streamt>(new json_stream_objectt(out, indent + 1));
+  return static_cast<json_stream_objectt &>(*current);
+}
```
This code is really hard to read: lots of bits and pieces that don't seem to fit together in any obvious way.

peterschrammel replied:
I agree. There is some doxygen work to be done to explain how things fit together.

tautschnig commented on src/util/json_stream.cpp:
```diff
@@ -0,0 +1,142 @@
+  return create_current_array();
+}
+
+/// Overloaded version of the push_back_stream_object, that
```
Nit pick: no comma before "that"

tautschnig commented on src/util/json_stream.h:
```diff
@@ -0,0 +1,151 @@
+/// This approach was necessary to fix some issues with
+/// CBMC running out of memory when it had to print some
+/// huge traces for some programs.
+class json_streamt
```
The only public operation is `close` - that's truly weird. Please create a meaningful API, which might have `virtual ...(...) = 0;` members - but at least it would explain how to use this thingy.

peterschrammel replied:
The only public facility that this base class provides is close().
output_current and output_finalizer should be made abstract.

tautschnig replied:
I don't think this has happened? Maybe it has, but they are all `protected` so I have no idea what is the public interface of (children of) `json_streamt`.

tautschnig commented on src/util/json_stream.h:
```diff
@@ -0,0 +1,151 @@
+  /// JSON object.
+  /// \param key: The key to be looked up inside the 
+  ///   attributes of the JSON object.
+  jsont &operator[](const std::string &key)
```
How does this fit in a "stream" class?!

peterschrammel replied:
It allows mixing streamed and non-streamed json elements. The API enforces that the parent of a streamed element must always be a streamed element.

tautschnig replied:
Why does it even make sense to mix them? Or is my above thought that a `json_streamt` is actually also some form of `jsont` that actually holds data?!

tautschnig commented on src/cbmc/all_properties.cpp:
```diff
@@ -207,18 +207,18 @@ void bmc_all_propertiest::report(const cover_goalst &cover_goals)
     {
       json_objectt json_result;
-      json_arrayt &result_array=json_result["result"].make_array();
+      json_arrayt &result_array = json_result["result"].make_array();
```
Whitespace changes?

peterschrammel replied:
These whitespace changes are unnecessary.

tautschnig commented on src/goto-programs/json_goto_trace.h:
```diff
@@ -14,14 +14,140 @@ Date: November 2005
-  trace_optionst trace_options = trace_optionst::default_options);
+  const namespacet &ns,
+  const goto_tracet &goto_trace,
+  json_arrayT &dest_array,
```
Why does this need to be a template argument?

peterschrammel replied:
To be able to use both the streaming and the object version.

tautschnig reviewed (commented):
The PR talks about fixing an out-of-memory problem, but then in the middle of the PR suddenly there is a refactoring commit as well. Lots of seemingly unnecessary whitespace changes. All of that makes reviewing extremely hard.

My main concern, however, is the core API that is not at all self-documenting. I get the problem that it is trying to solve, but it's very hard to tell whether that solution is actually fit for purpose. See details in my comments.

NlightNFotis committed:
Introduce class for streaming JSON to the output

Introduce a new json object class that streams the object
immediately to an output instead of building the object
in memory and then output it to the stream.

NlightNFotis committed:
Base JSON UI message handler upon json_streamt

NlightNFotis committed:
Refactor huge convert function into multiple ones

Refactor the templated convert function with switch
statements that handle different instructions to dispatch
to functions that are specialised per instruction.

NlightNFotis committed:
Activate JSON streaming when printing goto-traces

peterschrammel commented on src/cbmc/bmc_cover.cpp:
```diff
@@ -369,7 +369,7 @@ bool bmc_covert::operator()()
         if(bmc.options.get_bool_option("trace"))
         {
-          jsont &json_trace=result["trace"];
+          jsont &json_trace = result["trace"];
```
unnecessary.

peterschrammel commented on src/goto-programs/json_goto_trace.cpp:
```diff
@@ -33,20 +33,20 @@ void convert(
   trace_optionst trace_options)
 {
-  json_arrayt &dest_array=dest.make_array();
+  json_arrayt &dest_array = dest.make_array();
```
The changes in this file are unnecessary.

peterschrammel commented on src/util/json_stream.h:
```diff
@@ -0,0 +1,150 @@
+class json_stream_objectt;
+class json_stream_arrayt;
+
+/// New class whose aim is to stream newly created JSON
```
This "new class" will soon be an "old class" ;)

peterschrammel commented on src/util/message.h:
```diff
@@ -209,7 +209,17 @@ class messaget
     }
 
-    mstreamt &operator << (const json_objectt &data)
+    mstreamt &operator<<(const jsont &data)
```
This change seems unnecessary.

peterschrammel commented on src/util/json_stream.cpp:
```diff
@@ -0,0 +1,153 @@
+  out << '[';
+}
+
+/// Output the non-streaming Flushes and closes the current child stream
```
This comment looks broken

peterschrammel commented on src/goto-programs/json_goto_trace.h:
```diff
@@ -14,14 +14,145 @@ Date: November 2005
+} conversion_dependenciest;
+
+/// Convert an ASSERT step.
+void convert_assert(
```
\param docs missing for these functions

peterschrammel commented on src/goto-programs/json_goto_trace.h:
```diff
@@ -14,14 +14,145 @@ Date: November 2005
 #include <util/json.h>
+#include <util/json_stream.h>
+#include <util/json_expr.h>
+#include <util/arith_tools.h>
```
unused here, move to .cpp file

peterschrammel commented on src/goto-programs/json_goto_trace.h:
```diff
@@ -14,14 +14,145 @@ Date: November 2005
+#include <util/json_stream.h>
+#include <util/json_expr.h>
+#include <util/arith_tools.h>
+#include <util/config.h>
```
unused here, move to .cpp file

peterschrammel commented on src/goto-programs/json_goto_trace.h:
```diff
@@ -14,14 +14,145 @@ Date: November 2005
+#include <util/arith_tools.h>
+#include <util/config.h>
+#include <util/invariant.h>
+#include <util/simplify_expr.h>
```
unused here, move to .cpp file

peterschrammel commented on src/goto-programs/json_goto_trace.h:
```diff
@@ -14,14 +14,145 @@ Date: November 2005
+#include <util/config.h>
+#include <util/invariant.h>
+#include <util/simplify_expr.h>
+#include <util/json_irep.h>
```
unused here, move to .cpp file

peterschrammel commented on src/goto-programs/json_goto_trace.h:
```diff
@@ -14,14 +14,145 @@ Date: November 2005
+#include <util/invariant.h>
+#include <util/simplify_expr.h>
+#include <util/json_irep.h>
+#include <langapi/language_util.h>
```
unused here, move to .cpp file

peterschrammel reviewed (changes requested):
Commit 2 (10204a68) still contains changes that must be squashed into commit 1 or 3.

peterschrammel commented on src/goto-programs/json_goto_trace.h:
```diff
@@ -14,14 +14,178 @@ Date: November 2005
+      }
+    }
+
+    if(source_location.is_not_nil() && source_location.get_file() != "")
```
! .empty()

peterschrammel commented on src/goto-programs/json_goto_trace.h:
```diff
@@ -14,14 +14,178 @@ Date: November 2005
+
+    jsont json_location;
+
+    source_location.is_not_nil() && source_location.get_file() != ""
```
! .empty()

peterschrammel reviewed (commented):
@NlightNFotis, please create TG bump for this.

NlightNFotis commented:
I have added one last commit that attempts to fix the `goto-diff` broken tests, due to the way they interact with the new class.

I will write a lengthier report about what was happening and why this is necessary later.

NlightNFotis committed:
Fix goto-diff tests that got broken by streaming json

peterschrammel reviewed (approved):
Thanks, @NlightNFotis, this looks good now. Please update the TG bump.

tautschnig commented on src/util/message.h:
```diff
@@ -33,6 +34,12 @@ class message_handlert
+  virtual json_stream_arrayt &get_json_stream()
+  {
+    UNREACHABLE;
+  }
```
Please explain why this isn't `get_json_stream() = 0;` or marked protected?

NlightNFotis replied:
If I change it to `get_json_stream() = 0;` then I get compilation errors that some variables, such as  `message_handler` in line 189 of `graphml.cpp` can not be initialised to an abstract type. I understand that this happens because other members of the hierarchy, like `null_message_handlert` are not redefining `get_json_stream()`, a now pure virtual function.

If it's marked `protected`, that creates problems with the interaction with the class `mstreamt` in the same file (`message.h`, line 184), particularly with the method `json_stream()` in line 244, which contains the statement `return message.message_handler->get_json_stream();`, which can not work with `get_json_stream()` being protected.

tautschnig replied:
> If I change it to `get_json_stream() = 0;` then I get compilation errors that some variables, such as `message_handler` in line 189 of `graphml.cpp` can not be initialised to an abstract type. I understand that this happens because other members of the hierarchy, like `null_message_handlert` are not redefining `get_json_stream()`, a now pure virtual function.

To me, this looks like a prime example of why compile-time errors are preferable over runtime errors: in my opinion, that just means that `null_message_handlert` needs to be fixed.

tautschnig commented on src/util/message.h:
```diff
@@ -16,6 +16,7 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "invariant.h"
 #include "json.h"
+#include "json_stream.h"
```
I believe a forward declaration `class json_stream_arrayt;` would suffice.

tautschnig commented on src/util/json.h:
```diff
@@ -106,7 +106,6 @@ class jsont
       return it->second;
   }
 
-protected:
```
If you remove this, then also the `public` in line 123 is redundant.

peterschrammel replied:
@NlightNFotis, this shouldn't be removed. Only move the function that is required to the public section.

tautschnig commented on src/util/json_stream.cpp:
```diff
@@ -0,0 +1,153 @@
+  if(child_stream)
+  {
+    child_stream->close();
+  }
```
Isn't this prone to attempts of closing a closed file? `child_stream` is not reset.

peterschrammel replied:
Closing closed files is not a practical issue here because json_streamt only wraps the actual stream (that could be a file). json_streamt never closes that stream - that's in the realm of the user of json_streamt. However, to avoid confusion, I would assign child_stream = nullptr here and put appropriate PRECONDITIONS to make sure that there is no double-close (which is indeed a sign of a programming error).

tautschnig commented on src/util/json_stream.cpp:
```diff
@@ -0,0 +1,153 @@
+  output_delimiter();
+  jsont::output_key(out, key);
+  return create_child_stream_object();
+}
```
What is the purpose of all these `push_back_stream*` functions ("Add a JSON ..." -- to what?) as none of them are invoked anywhere in this commit, and why aren't they just a single one in the parent class? I understand that they have different return types, but is that actually necessary (as anything done to those return values can be resolved to the right child class via virtual tables)?

peterschrammel replied:
It would be very inconvenient to let the user cast to json_stream_arrayt and json_stream_objectt in order to use their respective interfaces (objects and arrays are quite different things).

tautschnig replied:
I think this fits with my overall comment: there seems to be a strong case for re-architecting here to use a visitor rather than this proposed design of inheritance.

peterschrammel replied:
@tautschnig, could you please explain how 'use a visitor' would contribute to a solution in this context?
Inheritance is used here because json_stream_arrayt and json_stream_objectt are both json streams, which share common functionality, namely flushing the data. If this is considered too little to justify a base class the common code can be duplicated in the respective derived classes at the cost of complicating the logic to flush the stream and to enforce the single-stream-at-a-time invariant.

tautschnig replied:
In my opinion, there should be one class that knows how to stream `jsont`. That class would be the visitor. It has methods that know how to stream a `json_arrayt` or `json_objectt` (and `json_numbert`, `json_stringt`, etc.).

tautschnig commented on src/util/json_stream.cpp:
```diff
@@ -0,0 +1,153 @@
+    output_delimiter();
+    object["array_element"].output_rec(out, indent + 1);
+    object.clear();
+  }
```
Noob question: why does this look quite different from `json_stream_objectt::output_child_stream`?

peterschrammel replied:
They are actually quite similar. The only difference is that array elements don't have keys.

tautschnig commented on src/util/json_stream.h:
```diff
@@ -0,0 +1,174 @@
+    {
+      output_child_stream();
+      output_finalizer();
+      open = false;
```
Wouldn't it be good enough to use the `child_stream` as an indicator (resetting it after it is closed)?

peterschrammel replied:
No. `open` is the state of `this` not `child_stream`. You can create and close many child streams sequentially (timewise), e.g. for each element in an array or each property in an object). The invariant is that there can only be at most child stream at a time.

tautschnig commented on src/util/json_stream.h:
```diff
@@ -0,0 +1,174 @@
+  // To be overridden by derived classes:
+  virtual void output_child_stream() = 0;
+  virtual void output_finalizer() = 0;
+};
```
Destructor to be marked virtual?

tautschnig commented on src/util/json_stream.h:
```diff
@@ -0,0 +1,174 @@
+  /// array stream.
+  /// Provided for compatibility with `jsont`.
+  /// \return a new empty, non-streaming JSON object
+  jsont &push_back()
```
Why do you need compatibility with `jsont`? Are you saying that this some sort of replacement for `jsont`? Please tell me that I'm wrong.

tautschnig commented on src/util/ui_message.h:
```diff
@@ -10,9 +10,14 @@ Author: Daniel Kroening, kroening@kroening.com
 #define CPROVER_UTIL_UI_MESSAGE_H
 
+#include <memory>
+#include <iostream>
```
Nope. Not in a header file.

peterschrammel replied:
@NlightNFotis, the implementation of the constructor must go into the .cpp file.

tautschnig commented on src/util/ui_message.h:
```diff
@@ -26,7 +31,10 @@ class ui_message_handlert : public message_handlert
-    : _ui(uit::PLAIN), time(timestampert::make(timestampert::clockt::NONE))
+    : _ui(uit::PLAIN),
+      time(timestampert::make(timestampert::clockt::NONE)),
+      out(std::cout),
```
First, there seems to be no point of even having a `std::ostream` member as there is a single function that relies on it, and no way to setting it to any other value. Second, this has no place in a header file.

peterschrammel replied:
No, `out` is used many, many times in the .cpp file. Eventually, assigning `out` should be the only use of `std::cout` in the codebase.
The implementation of the constructor must be moved back into the cpp file.

tautschnig reviewed (changes requested):
Various comments below, with the definitive blocker of using `std::cout` (and thus iostream) in a header file.

I'm still not sure I'm on the right track: do we now have multiple classes that serve one and the same purpose, but differ in how output is implemented? Does this mix data representation and behaviour? Are you solving an original design deficiency in that a `jsont` took care of its own output by adding a variant that takes care of its own output in a different way? Shouldn't `output` just never have been a member function of `jsont`? Isn't this one of the prototypical situations to use the visitor pattern? Even more so as `jsont` is the root of a nice family of classes, and also a very much bounded one so that it wouldn't be necessary to add more members to a visitor (as is sort of the case for anything based on `irept`)?

peterschrammel commented on src/util/json_stream.h:
```diff
@@ -0,0 +1,174 @@
+  {
+  }
+
+  bool open;
```
@NlightNFotis, it seems that `open` requires explanation.

NlightNFotis commented:
Hi @tautschnig. I think your concerns have been addressed. Is it possible that you take another look and let me know if there's anything more you want to be done, else give it the green light?

Thanks for your time,

tautschnig commented:
To what extent has https://github.com/diffblue/cbmc/pull/1814#discussion_r170541499:
>  think this fits with my overall comment: there seems to be a strong case for re-architecting here to use a visitor rather than this proposed design of inheritance.

been addressed? Anyone is free to override my non-approval, but I do maintain that the architecture here is flawed and that implementing a different form of output by inheriting and overriding the data representation is wrong.

martin-cs commented:
@tautschnig As the person who had the initial problem -- would it help if I had a review?

tautschnig commented:
@martin-cs Your reviews are always welcome! It may very well be the case that an urgent solution is required and I thus invite overrides to my non-approval. Any maybe my suggested architecture is also flawed. I'm always open to input.

NlightNFotis commented:
@tautschnig @peterschrammel I have been made aware that there have been offline discussions regarding the architectural changes in this PR, and the disagreements have been resolved. Can you both take a look to let me know if everything is alright, and if yes, possibly merge it as well?

@peterschrammel The associated testgen PR is also passing with no problems

tautschnig commented:
My apologies for having been dense and not fully understanding the problem this was trying to solve. It took an out-of-band chat with @peterschrammel to get me on track: it was not clear to me that constructing the `jsont` (and keeping it in memory) was the problem, not just constructing a very long `std::string` of output. With that insight I was able to make sense of all the bits that make up the proposed API. It's not the way I would have solved the problem (see below), but it's an existing solution so I'm not going to stand in the way any further.

Here is what I would have done: make `json_streamt` a factory (of sorts) that owns all `jsont` objects and can memory-manage them. Rather than changing `jsont` (or its children) I'd keep a stack of "open" objects in `json_streamt`, with any `push_back` or `operator[]` access going through the factory. The desired advantage is that the data representation would remain with `jsont`, and it would seem closer to an established design pattern. But it might have disadvantages that one would only get to see when really implementing it in full.
